### ISSUE TITLE
Figure out how we want to test gateway

### ISSUE THREAD

**Issue by jsam** [zeebe-io/zeebe#1124]
With the current setup, there is a lot of mocking involved. Maybe we can find a better strategy to do/improve testing there.

AC

* Come up with a strategy on how to test all endpoints on the gateway
* Current tests are adjusted to this strategy

**Comment by menski** [comment]
@jsam I was playing around with sending the requests async from the gateway to the cluster (251c56dcf24ea48d7ecbdd1b831bc4101beebe02). I therefore had to change some API and test setup cause I now uses the actor scheduler. While doing so I might have found a way how we can eliminate the client API mocking orgy and reduce the needed mocking for tests to a minimum. I is similar to my initial proposal to encapsulate the grpc request to cluster response mapping. This way we can ignore the ZeebeClient API and just mock the single call to the ClusterClient.

https://github.com/zeebe-io/zeebe/blob/251c56dcf24ea48d7ecbdd1b831bc4101beebe02/gateway/src/main/java/io/zeebe/gateway/ClusterClient.java#L30-L32

I think the endpoint test cases now boil down to a simple pattern. Maybe I can walk you through tomorrow if you are interested.

https://github.com/zeebe-io/zeebe/blob/251c56dcf24ea48d7ecbdd1b831bc4101beebe02/gateway/src/test/java/io/zeebe/gateway/HealthCheckEndpointTest.java#L77-L102

**Comment by jsam** [comment]
Great work! Thanks for figuring this out.

Can you explain to me though why this sudden change of heart? I thought we made a conscious decision that we will push this topic for later on. Why sudden urgency and change of the decision? Is there anything blocking or preventing us from doing for this sudden urgency now?

Would appreciate that we make decisions in future which actually matter and communicate that more clearly.

**Comment by menski** [comment]
> I thought we made a conscious decision that we will push this topic for later on.

I'm not hundert percent sure which topic you mean. But if it is about the async requests from gateway to broker then the answer is that nothing changed in our decision. As I'm currently not able to work on a topic focused over a longer period of time I just sometimes grab issue which are not urgent and maybe can be finished by me in couple of days, at least as PoC.

My work on the #1164 issue is just a start of the setup to figure out how this could look like. And while playing around with it, I saw that this approach of encapsulating the gateway2broker client could save us some work in the testing topic you are working on so I wanted to share this with you. It's still your decision if you want to pursue this direction at the moment or want to continue with the mocking approach you validated.

We can talk about this in more detail if you are interested. My comment wasn't meant to force a change in any decision we made.
